### DISCUSSION TITLE
Synchronizing two camera + audio

### DISCUSSION THREAD

**Discussion by YUNoSol** [bonsai-rx/bonsai#1662] [Q&A] [answered]
Hi.

Fairly new to Bonsai, and looking forward to using it in a new project. 

I wish to capture video from two cameras, and audio at the same time.

I have managed to combine my two videos into one, but there is a difference between the length of my video and audio recording. The difference depends on the time that I am recording, but the audio is always shorter than the video. a few examples: 
Video:4 sec, audio: 3 sec
Video:15 sec, audio: 13 sec
Video:6 sec, audio: 4 sec

I noticed a similar issue here: https://groups.google.com/g/bonsai-users/c/_ZAKxBot3iw 
I tried using the Slice function, where It got better, but still with a few seconds difference (on 30s+ videos) but I could never get a perfect match between the two recordings. 
I have tried changing the FPS for CameraCapture+VideoWriter, and the sample rate for AudioWriter and AudioCapture.

Here is an image of my workflow ( without the slicer, which goes in between Item1 and AudioWriter, like suggested in the issue on the google groups) 
![image](https://github.com/bonsai-rx/bonsai/assets/157392652/9bb84b10-a185-44e0-a30b-a266407d2847)

Does anyone have an idea of what I might be doing wrong?

-Sol

**Comment by YUNoSol** [comment]
Here is the workflow
[Bonsai-record-two-videos-with-sound-seperatly.zip](https://github.com/bonsai-rx/bonsai/files/14012280/Bonsai-record-two-videos-with-sound-seperatly.zip)

**Reply by bruno-f-cruz** [reply, marked as answer]
Welcome to the Discussions!

How are you determining the length of the video? I dont see how they would ever have the same duration (given by frames * encoded fps). The mic is acquiring at 44100(fps)/10(buffer size) = 4410 buffers per second. Your camera is acquiring at an unkown frequency but you force, with `WithLatestFrom` operator a sampling from the camera stream at 4410 frames per second (this means you are duplicating several the same frame >1000 times likely). Since you have a 4410 encoded at 30fps (VideoWriter setting) you will end up with a very long clip for sure. Not sure what the best way to go around this would be, but i would intuitively try to match, as much as possible, the acquisition rate of the sound buffers and the camera. For instance, if your camera acquires at 30fps, make your buffer size = 44100/30 = 1470 samples. This should at least give you a roughly matched video and sound clip length.

Hope it helps!

**Reply by YUNoSol** [reply]
Thank you Bruno!

I'm determining the length by opening the avi and wav files in programs like adobe premiere pro or windows media player.

I checked my webcams, and they should both acquire at 30 fps (logilink ua0377, and sandberg usb webcam 1080p HD).- I am aware that it would be better to record with similar webcam, that is a thing we are looking into. 

I tried your suggestion, and changed the buffer lenght within AudioWriter, with no luck. I tried increasing and decreasing the BufferLenght to get as close as possible, but the closest I can get is still missing more than a second. I tried with 29fps within VideoWriter, in case the cameras do not live up the 30 fps they claim, but this also only got me to a second missing (I wrote down the combinations I tried, see below). 

_With 44100(Audiocapture and AudioWriter), and 30 fps(VideoWriter):
Attempting to change BufferLenght within AudioCapture.

Buffer to 30= 18 sec video, and 16 sec audio. 
Buffer to 29 =15s audio, 18 s video
Buffer to 32= Audio 1 min, 13 sec. video 1 min 16 sec. 
Buffer to 34=Audio 1 min 41 sec.. Video 1 min 39 sec
Buffer to 33=Audio 2 min 39 sec. Video 2 min 36 sec. 
Attempt to change VideoWriter setting capture to 29 fps:
Buffer to 29.= Audio 56 sec. Video 1 min 7 sec
Buffer 30= Audio 2 min 46 se. video 3 min 10 sec. 
Buffer to 33= Audio 2 min 14 sec video 2 min 20 sec. 
Buffer 34= Audio 19 sec video 20 sec.
Buffer 35= Audio 53 s, video 52 s. -> view in adobe premiere pro. Here audio 52.20, and video 51.26._


If you or anyone else has another Idea, I would be grateful, worst case scenario I'll have to attempt to stretch the audio to match the video every time, but I of course prefer to synchronize instead. 

-Sol

**Reply by bruno-f-cruz** [reply]
The last test is pretty good. Out of curiosity,if you add a zip between the two final sink nodes, does it get better? 

Also if you want to correct you might be able to get away with just modifying the metadata of the video and change the playback frame rate.

**Reply by YUNoSol** [reply]
Hi Again

I tried to put a zip between the two final nodes as shown below ( this was the only way it could get it to work), but it did not help unfortunately. ( Tried longer videos, so now I get 6 min 52 audio, 6 min 46 video,  and 7 min 19 video with 7 min 27 audio.

![image](https://github.com/bonsai-rx/bonsai/assets/157392652/5fb6f6a4-22bb-4c3c-abfb-f38b062ac82d)

I'm not sure about how to change the metadata, but the best alternative I have found yet, was to change the speed of the audio, to match the duration of the video, within Premiere Pro, which seems to match almost perfectly (few frames off, which might be due to camera fps, and that I'm still working with two different cameras).

Thank you for the help anyways!

-Sol

**Reply by bruno-f-cruz** [reply]
Add the zip to prevent dangling branches.

![image](https://github.com/bonsai-rx/bonsai/assets/7049351/b876f289-1b8b-4aa6-b5b9-893f69a920ef)

**Reply by YUNoSol** [reply]
Thank you! I misunderstood you firstly, that makes more sense.

I'm still missing a few seconds, but that can be fixed in an editing software as mentioned before. 

This is my final workflow, with timestamps and KeyDown incorporated, for anyone wanting to do something simillar. 

![image](https://github.com/bonsai-rx/bonsai/assets/157392652/2087c06c-e972-4a5d-bf83-63a24bfde371)

**Reply by bruno-f-cruz** [reply]
That s odd.. can you attach your workflow?

**Reply by YUNoSol** [reply]
Of course! 

Here is the workflow
[ForUpload.zip](https://github.com/bonsai-rx/bonsai/files/14039726/ForUpload.zip)

And here are my outputs.
The difference here is only video=20.24 and audio 21.03 when viewing in premiere pro.
[output.zip](https://github.com/bonsai-rx/bonsai/files/14039857/output.zip)

**Reply by bruno-f-cruz** [reply]
If I set `BufferLength` to 40milliseconds and `VideoWriter` to a corresponding 25fps I get something along the lines of what I would expect:

```cmd
ffmpeg -i timesweep.wav -f null -
ffmpeg -i timesweep.avi -f null -
```
return, respectively:

```cmd
(...) size=N/A time=00:01:19.41 (...)
(...) time=00:01:19.40 (...)
```
Not sure where those extra 10ms come from honestly, but it might have something to do with how time is rounded for video vs sound. Generally I would match number of samples not time, as the latter relies on the container metadata (i.e. frames per second)

**Reply by YUNoSol** [reply]
This works perfectly!
Thank you so much for all your help!
